### PR TITLE
Add Fastly log processor

### PR DESCRIPTION
**Pull request by ktheory**
:wave: Happy Friday :ok_woman: 

This implements much of rubygems/rubygems-infrastructure#35, allowing us to update Rubygem download counts in bulk using Fastly access logs rather than on each HTTP request.

I still have some cleanup to do, but wanted to get some :eyes: earlier.

### To Do
- [x] add a feature flag to control whether to have FastlyLogProcessor actually update stats, or just log results. (It would only log initially.)
- [ ] figure out how to disable the `stat-update` C library as part of the feature flag
- [x] Cleanup & test shoryuken config, preferably loading SQS queue names from ENV vars
- [x] Make another PR to run a `shoryuken` process on deploy once Chef provisions a runit script for it

### Deploy plan

NB that this code is disabled right now. We're not currently running a `shoryuken` worker, and so nothing would enqueue a `FastlyLogProcessor` job.

Once this is merged, the next step is to start running a `shoryuken` process, with `FastlyLogProcessor` only logging results & not updating redis.

After that I'll toggle the feature flag so `FastlyLogProcessor` updates stats, and disable `stat-update` per request.

FYI @dwradcliffe

### CODE REVIEW

**Comment by ktheory** [comment]
Rubocop is not a happy. :scream:

I'll amend the commits. :smile:

File: config/initializers/delayed_job.rb [hunk 1]
```diff
@@ -3,4 +3,4 @@
 Delayed::Worker.max_run_time = 5.minutes
 Delayed::Worker.logger = Logger.new(Rails.root.join('log/delayed_job.log'))
 
-PRIORITIES = { push: 1, download: 2, web_hook: 3 }.freeze
+PRIORITIES = { push: 1, download: 2, web_hook: 3, download_metrics: 4 }.freeze
```
Comment (dwradcliffe) [inline config/initializers/delayed_job.rb]: Can we shorten this to `stats`?

File: Gemfile.lock [hunk 2]
```diff
@@ -65,6 +69,23 @@ GEM
...
+    celluloid-essentials (0.20.5)
+      timers (>= 4.1.1)
+    celluloid-extras (0.20.5)
+      timers (>= 4.1.1)
+    celluloid-fsm (0.20.5)
+      timers (>= 4.1.1)
+    celluloid-pool (0.20.5)
+      timers (>= 4.1.1)
+    celluloid-supervision (0.20.5)
+      timers (>= 4.1.1)
```
Comment (dwradcliffe) [inline Gemfile.lock]: Too bad we need all the celluloid gems :/

File: app/jobs/fastly_log_processor.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  # E.g.
+  #   {
+  #     'rails-4.0.0' => 25,
+  #     'rails-4.2.0' => 50
+  #   }
+  def download_counts(enumerator = log_lines)
+    enumerator.each_with_object(Hash.new(0)) do |log_line, accum|
+      path, response_code = log_line.split[10, 2]
+      # Only count successful downloads
+      # TODO: should we count 304 not modified responses?
```
Comment (dwradcliffe) [inline app/jobs/fastly_log_processor.rb]: A 304 is a download attempt, so I think that counts? Right now they would be real downloads so we should probably keep counting them. Any other opinions?

**Comment by dwradcliffe** [comment]
I added a few comments and it looks like this needs a rebase.

File: app/jobs/sqs_worker.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+class SqsWorker
+  include Shoryuken::Worker
+
+  # TODO: set real queue name
+  # TODO: set auto_delete: true after testing
+  shoryuken_options queue: 'TODO-add-real-queue', body_parser: :json, auto_delete: false
```
Comment (qrush) [inline app/jobs/sqs_worker.rb]: Do we actually need Shoryuken here? What's the advantage to using this when we have Delayed::Job?

File: app/models/download.rb [hunk 5]
```diff
@@ -2,15 +2,31 @@ class Download
...
-    Redis.current.hincrby(rubygem_history_key(name), today, 1)
+    redis = Redis.current
+
+    redis.incrby(COUNT_KEY, count)
+    redis.incrby(rubygem_key(name), count)
+    redis.incrby(version_key(full_name), count)
+    redis.zincrby(today_key, count, full_name)
+    redis.zincrby(ALL_KEY, count, full_name)
+    redis.hincrby(version_history_key(full_name), today, count)
+    redis.hincrby(rubygem_history_key(name), today, count)
```
Comment (qrush) [inline app/models/download.rb]: Could wrap all of these in a `redis.pipelined` block, it'll be a lot faster since we don't need to wait for any replies.

File: app/models/download.rb [hunk 6]
```diff
@@ -2,15 +2,31 @@ class Download
...
+  end
+
+  # Takes an array where members have the form
+  #   [name, full_name, count]
+  # E.g.:
+  #   ['rake', 'rake-10.4.2', 1]
+  def self.bulk_update(ary)
+    Redis.current.multi do
```
Comment (qrush) [inline app/models/download.rb]: Why do we need a `MULTI/EXEC` transaction here? I think pipelining is probably fast enough. Are we worried some counts might be inaccurate?

File: script/shoryuken [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+exec bin/shoryuken --rails -r app/jobs/sqs_worker.rb
```
Comment (qrush) [inline script/shoryuken]: This should go in `bin/` now

File: test/unit/fastly_log_processor_test.rb [hunk 8]
```diff
@@ -0,0 +1,99 @@
+require 'test_helper'
+
+class FastlyLogProcessorTest < ActiveSupport::TestCase
+  setup do
+    @sample_log = File.read(Rails.root.join('test/sample_logs/fastly-fake.log'))
```
Comment (qrush) [inline test/unit/fastly_log_processor_test.rb]: Can use `Pathname#read` here instead:

`Rails.root.join('test/sample_logs/fastly-fake.log').read`

File: app/jobs/fastly_log_processor.rb (hunk 3 above)
Comment (qrush) [inline app/jobs/fastly_log_processor.rb]: I'd say yes, count any attempt.

File: app/jobs/fastly_log_processor.rb [hunk 9]
```diff
@@ -0,0 +1,75 @@
+require 'zlib'
+FastlyLogProcessor = Struct.new(:bucket, :key) do
+  def perform
+    counts = download_counts
+
+    unless Gemcutter::ENABLE_FASTLY_LOG_PROCESSOR
```
Comment (qrush) [inline app/jobs/fastly_log_processor.rb]: This seems like a weird place for the feature flag. Can we move it to where we enqueue the job instead?

File: Gemfile.lock (hunk 2 above)
Comment (arthurnn) [inline Gemfile.lock]: do we? why?

Comment (arthurnn) [inline Gemfile.lock]: `shoryuken` gem. right.. I guess it is fine for now. I can try to remove the dependency afterwards, and write something simpler.

File: config/initializers/delayed_job.rb (hunk 1 above)
Comment (arthurnn) [inline config/initializers/delayed_job.rb]: I hate this constant been here :( eventually i wanna moved it out. I had that somewhere. Anyways, not in the scope of this PR

File: config/initializers/download_processor.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+module Gemcutter
+  # Feature flag for counting downloads through stats-update or FastlyLogProcessor
+  ENABLE_FASTLY_LOG_PROCESSOR = Rails.env.test? || Rails.env.development?
```
Comment (arthurnn) [inline config/initializers/download_processor.rb]: why do we need this constant here? can we move this to the Download model or the log processor itself?

File: app/jobs/fastly_log_processor.rb [hunk 11]
```diff
@@ -0,0 +1,75 @@
+require 'zlib'
+FastlyLogProcessor = Struct.new(:bucket, :key) do
```
Comment (arthurnn) [inline app/jobs/fastly_log_processor.rb]: I dont like Structs. they are usually slower than normal classes too. Can we define a class here instead?

Comment (arthurnn) [inline app/jobs/fastly_log_processor.rb]: cc @kerrizor :)

Comment (ktheory) [inline app/jobs/fastly_log_processor.rb]: > I dont like Structs. they are usually slower than normal classes too. Can we define a class here instead?

I was following app/jobs/notifier.rb as an example (which uses a Struct), and the Struct saves me some boilerplate initializer/accessor code.

I was aiming to be as conventional as possible. :smile: @arthurnn Is the project's convention now to use classes?

File: config/initializers/download_processor.rb (hunk 10 above)
Comment (ktheory) [inline config/initializers/download_processor.rb]: > why do we need this constant here? can we move this to the Download model or the log processor itself?

We don't need it *here*, but we need it somewhere. :smile:

Since it's serving as a feature flag, it's more appropriate as config than as app or library code. My rule of thumb is that app code can talk to config code; and config code can use `Rails.env`. But app code should never directly check `Rails.env`.

`app <--> config <--> environments`

FWIW, this is a temporary feature flag while we dark-launch Fastly log processing. Once we're confident it's working well, we'd remove this.

File: app/jobs/fastly_log_processor.rb (hunk 9 above)
Comment (ktheory) [inline app/jobs/fastly_log_processor.rb]: @qrush We still want to enqueue the job & process the log regardless.

This feature flag is toggling whether we actually update the counts in redis; or just log the counts for debugging.

NB this is a temporary feature flag while we dark-launch log processing. I'll remove it once we're confident it's working well.

File: app/models/download.rb (hunk 5 above)
Comment (ktheory) [inline app/models/download.rb]: @qrush:

> Could wrap all of these in a redis.pipelined block, it'll be a lot faster since we don't need to wait for any replies.

The `bulk_update` method that calls this already sets up a pipeline. :smile:

File: app/models/download.rb (hunk 6 above)
Comment (ktheory) [inline app/models/download.rb]: @qrush:
> Why do we need a MULTI/EXEC transaction here?

It's for data consistency.

If a ruby process crashes part-way through updating a bunch of stats, the transaction will rollback, and the FastlyLogProcessor job will eventually retry.

Either *all* the redis commands succeed and FastlyLogProcessor job succeeds. Or the whole transaction fails, and the FastlyLogProcessor job retries.

File: app/jobs/sqs_worker.rb (hunk 4 above)
Comment (ktheory) [inline app/jobs/sqs_worker.rb]: @qrush:
> Do we actually need Shoryuken here? What's the advantage to using this when we have Delayed::Job?

We need a tool that uses AWS SQS as the queue. See [this comment & discussion on rubygems-infrastructure](https://github.com/rubygems/rubygems-infrastructure/issues/35#issuecomment-160700892).

File: app/jobs/fastly_log_processor.rb (hunk 11 above)
Comment (arthurnn) [inline app/jobs/fastly_log_processor.rb]: yeah. I guess classes are better. Lets use a class, and maybe we can re-work the other notifier to change for a class too in the future

Comment (kerrizor) [inline app/jobs/fastly_log_processor.rb]: I see you've made the change already, but for the curious, Structs are ~46% _slower_ than using a Class (at least, for simple getter/setter actions.) [Some benchmarking](https://gist.github.com/kerrizor/e961504d1fbd425bfe97)

File: script/shoryuken (hunk 7 above)
Comment (ktheory) [inline script/shoryuken]: @qrush:
> This should go in bin/ now

Since `bin` is gitignored, it seemed like a non-starter executables like this. Any suggestions?

Comment (dwradcliffe) [inline script/shoryuken]: I'd say remove from gitignore, as we should be placing things in bin/ now.

Comment (ktheory) [inline script/shoryuken]: :ok_woman: Separate PR incoming.

Comment (ktheory) [inline script/shoryuken]: Actually, just realized I could add this with `git add -f`. :cake:

Comment (ktheory) [inline script/shoryuken]: @qrush, @dwradcliffe Fixed!

**Comment by dwradcliffe** [comment]
LGTM :ship:
